import "ydb/library/actors/protos/actors.proto";
import "ydb/core/protos/tablet.proto";
import "ydb/core/protos/base.proto";
import "ydb/core/protos/tracing.proto";

package NKikimrTracing;
option java_package = "ru.yandex.kikimr.proto";

message TOnCancelTablet {
    optional fixed64 TimeStamp = 1;
    optional uint64 TabletID = 2;
    optional NKikimrTabletBase.TTabletTypes.EType TabletType = 3;
    optional uint32 TabletDeadReason = 4;
    optional uint32 SuggestedGeneration = 5;
    optional uint32 KnownGeneration = 6;
}

message TOnTabletBootstrap {
    optional fixed64 TimeStamp = 1;
    optional bool Leader = 2;
    optional uint32 SuggestedGeneration = 3;
    optional NActorsProto.TActorId StateStorageProxyID = 4;
}

message TOnHandleStateStorageInfoResolve {
    optional fixed64 TimeStamp = 1;
    optional uint32 KnownGeneration = 2;
    optional uint32 KnownStep = 3;
    optional uint32 SignatureSz = 4;
}

message TOnLockedInitializationPath {
    optional fixed64 TimeStamp = 1;
    optional uint32 KnownGeneration = 2;
    optional uint32 KnownStep = 3;
    optional uint32 SignatureSz = 4;
}

message TOnHandleStateStorageInfoLock {
    optional fixed64 TimeStamp = 1;
    optional uint32 KnownGeneration = 2;
    optional uint32 KnownStep = 3;
    optional uint32 SignatureSz = 4;
}

message TOnPromoteToCandidate {
    optional fixed64 TimeStamp = 1;
    optional uint32 KnownGeneration = 2;
}

message TOnTabletBlockBlobStorage {
    optional fixed64 TimeStamp = 1;
    optional NActorsProto.TActorId ReqBlockBlobStorageID = 2;
    optional uint32 KnownGeneration = 3;
}

message TOnTabletRebuildGraph {
    optional fixed64 TimeStamp = 1;
    optional NActorsProto.TActorId TabletReqRebuildGraphID = 2;
    optional TTraceID RebuildGraphTraceID = 3;
}

message TOnWriteZeroEntry {
    optional fixed64 TimeStamp = 1;
    optional uint64 Snapshot = 2;
    optional uint64 ZeroConfirmed = 3;
    optional uint32 LastInGeneration = 4;
}

message TOnFollowerPromoteToLeader {
    optional fixed64 TimeStamp = 1;
    optional uint32 SuggestedGeneration = 2;
    optional NActorsProto.TActorId KnownLeaderID = 3;
    optional NActorsProto.TActorId FollowerStStGuardian = 4;
}

message TRebuildGraphBootstrap {
    optional fixed64 TimeStamp = 1;
    optional uint32 BlockedGen = 2;
}

message TErrorEntryBeyondBlocked {
    optional fixed64 TimeStamp = 1;
    optional NKikimrProto.TLogoBlobID Latest= 2;
    optional uint32 BlockedGen = 3;
}

message TErrorUnknownStatus {
    optional fixed64 TimeStamp = 1;
    optional NKikimrProto.EReplyStatus Status = 2;
    optional string Reason = 3;
}

message TErrorParsingFromString {
    optional fixed64 TimeStamp = 1;
    optional NKikimrProto.TLogoBlobID BlobID = 2;
}

message TErrorSendRefsCheck {
    optional fixed64 TimeStamp = 1;
}

message TErrorRebuildGraph {
    optional fixed64 TimeStamp = 1;
    optional uint32 Generation = 2;
    optional uint32 Step = 3;
}

message TOnProcessKeyEntry {
    optional fixed64 TimeStamp = 1;
    optional NKikimrProto.TLogoBlobID LatestEntry = 2;
}

message TOnProcessZeroEntry {
    optional fixed64 TimeStamp = 1;
    optional uint32 Generation = 2;
    optional uint64 Snapshot = 3;
    optional uint64 Confirmed = 4;
}

message TOnProcessLogEntry {
    optional fixed64 TimeStamp = 1;
    optional NKikimrProto.TLogoBlobID BlobID = 2;
    optional uint64 Snapshot = 3;
    optional uint64 Confirmed = 4;
    repeated NKikimrProto.TLogoBlobID References = 5;
    repeated NKikimrProto.TLogoBlobID GcDiscovered = 6;
    repeated NKikimrProto.TLogoBlobID GcLeft = 7;
}

message TOnDiscoverRangeRequest {
    optional fixed64 TimeStamp = 1;
    optional uint32 Group = 2;
    optional NKikimrProto.TLogoBlobID From = 3;
    optional NKikimrProto.TLogoBlobID To = 4;
}

message TOnApplyDiscoveryRange {
    optional fixed64 TimeStamp = 1;
    optional uint32 Group = 2;
    optional NKikimrProto.TLogoBlobID From = 3;
    optional NKikimrProto.TLogoBlobID To = 4;
}

message TOnMakeHistory {
    optional fixed64 TimeStamp = 1;
    repeated NKikimrProto.TLogoBlobID RefsToCheck = 2;
}

message TOnCheckRefsGetResult {
    optional fixed64 TimeStamp = 1;
    optional uint32 ResponseSz = 2;
}

message TOnBuildHistoryGraph {
    message TEntry {
        optional uint32 Generation = 1;
        optional uint32 Step = 2;
        optional bool IsSnapshot = 3;
        repeated NKikimrProto.TLogoBlobID References = 4;
        repeated NKikimrProto.TLogoBlobID GcDiscovered = 5;
        repeated NKikimrProto.TLogoBlobID GcLeft = 6;
    }
    
    optional fixed64 TimeStamp = 1;
    repeated TEntry Entries = 2;
}

message TOnRebuildGraphResult {
    optional fixed64 TimeStamp = 1;
    optional string SerializedTrace = 2;
}
